File: java/inheritance_by_class.java
import java.util.ArrayList;

public class Main {
  public static void main(String[] args) {
    ArrayList<Pet> petList = new ArrayList<Pet>();
    Dog dog = new Dog("Firulais");
    Cat cat = new Cat("Michifuz");
    petList.add(dog);
    petList.add(cat);
    for (Pet pet : petList) pet.animalSound();
    
  }
}

class Pet {
  protected String name;

  public Pet(String name) {
    this.name = name;
  }

  public void animalSound() {
    System.out.println("The animal makes a sound");
  }
}

class Dog extends Pet {
  public Dog(String name) {
    super(name);
  }

  public void animalSound() {;
    System.out.println("Guau");
  }
}

class Cat extends Pet {
  public Cat(String name) {
    super(name);
  }

  public void animalSound() {;
    System.out.println("Miau");
  }
}

File: c++/inheritance_by_class.cc
#include <iostream>
#include <string>
using namespace std;

class Pet {
  protected:
    string name;

  public: 
    Pet(string name) {
      name = name;
    }
    
    virtual void animalSound() {
      cout << "The animal makes a sound" << endl;
    }
};

class Dog : public Pet {
  public: 
    Dog(string nombre): Pet(nombre) {}

    void animalSound() {
      cout << "Guau" << endl;
    }
};

class Cat : public Pet {
  public: 
    Cat(string nombre): Pet(nombre) {}

    void animalSound() {
      cout << "Miau" << endl;
    }
};

int main() {
  Pet *petList[] = {new Dog("Firulais"), new Cat("Michifuz")};
  for (int i = 0; i < 2; i++) 
    petList[i]->animalSound();
  return 0;
}